Fix #4072 : Uses Initial few words as a feedback's subject

pandeykartikey opened the pull request:
Uses first few words as the subject of feedback instead of (Feedback from a learner).

pandeykartikey committed:
Fix #4072

pandeykartikey commented:
@seanlip @jacobdavis11 have a look

kevinlee12 commented on core/templates/dev/head/pages/exploration_player/FeedbackPopupDirective.js:
```diff
@@ -108,8 +108,9 @@ oppia.directive('feedbackPopup', [
 
           $scope.saveFeedback = function() {
             if ($scope.feedbackText) {
+              feedbackSubject = $scope.feedbackText.split(' ', 7).join(' '); 
```
This might be an issue with languages that don't have spaces and are character-based. Perhaps use a character count instead.

kevinlee12 replied:
and you'll need to declare a var here :)

pandeykartikey replied:
Done

kevinlee12 reviewed (changes requested):
Hi @kartik-25, let a comment, PTAL!

pandeykartikey committed:
Made required changes

pandeykartikey commented:
@kevinlee12 Made the required changes

kevinlee12 commented on core/templates/dev/head/pages/exploration_player/FeedbackPopupDirective.js:
```diff
@@ -108,8 +108,12 @@ oppia.directive('feedbackPopup', [
           $scope.saveFeedback = function() {
             if ($scope.feedbackText) {
+              var feedbackSubject = $scope.feedbackText.split(' ', 7).join(' ');
+              if (!feedbackSubject.includes(' ')) {
```
Not a fan of this check since languages such as Chinese do utilize spaces (but yet there are no spacing between characters) so it would result into the text being too long. Another strategy here would be to cut on character length first then by word.

pandeykartikey replied:
Done

kevinlee12 reviewed (changes requested):
Hi @kartik-25, it's heading in the right direction, but there is an unintended side effect of the change, please see my comment.

pandeykartikey committed:
Made required changes

pandeykartikey commented:
I first cut down by character and then if it contained spaces I extracted the words. I removed the last extracted element as it may be incomplete because of cut down by character. Hope this is fine @kevinlee12

kevinlee12 commented:
Hi @kartik-25, after some UI testing with lorem ipsum with Chinese, Arabic, and English, it seems that the character cut-off might be too generous, see:

![image](https://user-images.githubusercontent.com/5758256/34324687-63a1b4ee-e830-11e7-9fcc-449334de0af7.png)

I'd recommend tweaking the character cut-off so that the subject takes up at most one line. A good way to test is to use a lorem ipsum generator which generates realistic, but fake, text. (By the way, welcome to the messy world of internationalization :smile: )

pandeykartikey commented:
@kevinlee12 I have tweaked the character cut-off.

kevinlee12 commented:
Hmm, the ellipsis disappeared, this is what I see:
![image](https://user-images.githubusercontent.com/5758256/34324991-8f49f102-e839-11e7-8b22-0674c0a8c14e.png)

pandeykartikey commented:
ellipsis were there because I was exceeding the word-limit of the subject box

kevinlee12 commented:
Ah, in that case, I think going back to what you had before might better then since not having the ellipsis is worse from the user's perspective.

pandeykartikey commented:
reverted to the previous commit

kevinlee12 reviewed (dismissed):
lgtm, thanks @kartik-25! 

will merge when tests pass.

seanlip commented on core/templates/dev/head/pages/exploration_player/FeedbackPopupDirective.js:
```diff
@@ -108,8 +108,13 @@ oppia.directive('feedbackPopup', [
 
           $scope.saveFeedback = function() {
             if ($scope.feedbackText) {
+              var feedbackSubject = $scope.feedbackText.substr(0, 60);
```
This needs tests. Please pull it out into a separate function/service, something like getAbbreviatedFeedbackText(), and add Karma tests for it to ensure that the various cases are correctly handled.

pandeykartikey replied:
I don't understand why is there a need for tests in something like `substr()`?

seanlip replied:
Sorry, I don't mean just substr(). I mean the whole process of generating a subject string from a text message.

pandeykartikey replied:
@seanlip I have not written any tests before so could you guide me a bit here.

seanlip replied:
I suggest looking at the existing Karma tests, in files named *Spec.js. You'll want to extract this code into a function in its own service, or into a filter and then write tests for that function/filter.

@kevinlee12 should be able to provide more guidance, too, but you should be able to get started by looking at the existing code. There's also some guidance on [this wiki page](https://github.com/oppia/oppia/wiki/Running-Tests#client-side-tests).

seanlip reviewed (changes requested):
(Please see comment.)

seanlip commented:
Also, I don't see any functionality to add an ellipsis. Is this intentional?

pandeykartikey commented:
I thought it would not look nice when someone opens a feedback thread and ellipsis are shown in the title.

kevinlee12 commented:
Ellipsis would be needed otherwise the creator might think that that is shown is all what the feedback contains.

Just a side note, I think we should rename "Subject" to "Feedback" (or something similar) since it's more like a feedback preview than a title, what do you think @seanlip?

seanlip commented:
Hm, ok, this is raising some interesting questions.

It seems to me like this new meaning of "subject" isn't really so much a "subject" as it is a summary of the text of the first message in the feedback thread. In addition, it's possible for collaborators on an exploration to open feedback threads for each other. So it feels like we're overloading the "Subject" field here.

So, here's what I suggest instead: when retrieving the list of feedback threads from the backend, do an additional database call to get the content of the first message for each of those threads, and send this info over to the frontend. The frontend takes those messages, abbreviates them, and displays them. (It's also OK to do the abbreviation in the backend if you prefer.) That way, we don't change the meaning of the "subject" field midway (otherwise we'd have some model instances that say "Feedback from a learner" and some others that have abbreviated feedback text and some others that are handwritten by fellow creators).

What do y'all think?

kevinlee12 commented:
I'm happy with that.

seanlip commented:
Hi @kartik-25, friendly ping -- are you still working on this?

pandeykartikey commented:
Yes

pandeykartikey committed:
Moved to a separate function

pandeykartikey commented:
@kevinlee12 Can you give me some guidance what kind of things require tests?

kevinlee12 commented:
With your issue, you might want to check whether text under a certain length aren't cut, over they are cut, etc. I'd also recommend putting this in a service perhaps in `filters.js` and placing the tests in `filtersSpec.js`. 

Second opinion on whether `filters.js` is an appropriate place would be appreciated @seanlip.

seanlip commented:
@kevinlee12's suggestions SGTM (thanks!).

kevinlee12 commented:
Hi @kartik-25, how is this coming along?

pandeykartikey committed:
Added tests for Filter Service

pandeykartikey commented:
@kevinlee12 I have added the tests and shifted filter to a new service. Have a look.

kevinlee12 commented on core/templates/dev/head/services/FilterService.js:
```diff
@@ -0,0 +1,35 @@
+// limitations under the License.
+
+/**
+ * @fileoverview Service to serve as the interface for fetching and uploading
```
That's not the right description :)

pandeykartikey replied:
No need for description now 😄

kevinlee12 commented on core/templates/dev/head/services/FilterService.js:
```diff
@@ -0,0 +1,35 @@
+oppia.factory('FilterService', [function () {
+  return {
+    getAbbreviatedText: function (text) {
+      if (text.length > 50 ) {
```
Remove space before `)`

kevinlee12 replied:
It would be better if the limit were converted to a constant, since it minimizes any errors should we change it later on.

pandeykartikey replied:
Done

kevinlee12 commented on core/templates/dev/head/services/FilterService.js:
```diff
@@ -0,0 +1,35 @@
+  return {
+    getAbbreviatedText: function (text) {
+      if (text.length > 50 ) {
+        var subject = text.substr(0, 50);
```
Ditto

pandeykartikey replied:
Done

kevinlee12 commented on core/templates/dev/head/services/FilterServiceSpec.js:
```diff
@@ -0,0 +1,38 @@
+// limitations under the License.
+
+/**
+ * @fileoverview Unit tests for the Alerts Service.
```
Not the right description :)

pandeykartikey replied:
No need for description now 😄

kevinlee12 commented on core/templates/dev/head/services/FilterServiceSpec.js:
```diff
@@ -0,0 +1,38 @@
+  }));
+
+  it('should not shorten the length of text if less than 50', function() {
+    expect(FilterService.getAbbreviatedText('Lorem ipsum dolor'))
```
While it is fine to use lorem ipsum in manual tests, we try to avoid those in automated tests, so you'll want to use an actual sentence.

pandeykartikey replied:
Done

kevinlee12 commented on core/templates/dev/head/services/FilterServiceSpec.js:
```diff
@@ -0,0 +1,38 @@
+  });
+
+  it('should shorten the length of text if greater than 50', function() {
+    expect(FilterService.getAbbreviatedText(
```
Ditto

pandeykartikey replied:
Done

kevinlee12 commented on core/templates/dev/head/services/FilterServiceSpec.js:
```diff
@@ -0,0 +1,38 @@
+      'FilterService');
+  }));
+
+  it('should not shorten the length of text if less than 50', function() {
```
`....less than 50 characters.`

pandeykartikey replied:
Done

kevinlee12 commented on core/templates/dev/head/services/FilterServiceSpec.js:
```diff
@@ -0,0 +1,38 @@
+      .toBe('Lorem ipsum dolor');
+  });
+
+  it('should shorten the length of text if greater than 50', function() {
```
`...greater than 50 characters.`

pandeykartikey replied:
Done

kevinlee12 reviewed (dismissed):
Hi @kartik-25, looks like there's a misunderstanding regarding my comment. Since it is performing a filtering operation, it would be best if you could place these in `core/templates/dev/head/filters.js` and the tests in `core/templates/dev/head/filtersSpec.js`. I've also left some other comments at specific places. Please let me know if you have any questions! thanks!

pandeykartikey committed:
Made the required changes

pandeykartikey committed:
Merge branch 'develop' into feedback-title

pandeykartikey commented:
@kevinlee12 I have looked at all the comments and resolved the merge conflict. You can go on with the review.

kevinlee12 commented on core/templates/dev/head/filters.js:
```diff
@@ -407,6 +409,21 @@ oppia.filter('stripFormatting', [function() {
 }]);
 
+oppia.filter('getAbbreviatedText', [function () {
+  return function (text, characterCount) {
```
Is it possible to reference the limit directly here? It would reduce code complexity.

pandeykartikey replied:
@kevinlee12 If in future it is to be used somewhere else then it would be easier to specify character limit of that feature and use this function.

kevinlee12 commented:
@pranavsid98 will be taking over this PR since I'm under the weather.

pandeykartikey commented:
@pranavsid98 Any more changes required in this PR?

pranavsid98 commented:
@kartik-25 Oh, I just noticed the ping. Give me a day, I'll get my review in. Sorry!

pranavsid98 reviewed (approved):
@kartik-25 Seems Kevin and Sean covered the major concern points, everything currently looks great to me! There seems to be some issues with Travis e2e tests though (I restarted them once). If that passes, I think this is good to go!
Thanks!

/cc @seanlip

pandeykartikey commented:
@pranavsid98 Why is the document getting unloaded every time in tests?

seanlip commented on core/templates/dev/head/filters.js:
```diff
@@ -18,6 +18,8 @@
 
 oppia.constant('RULE_SUMMARY_WRAP_CHARACTER_COUNT', 30);
 
+oppia.constant('SUMMARRY_TITLE_CHARACTER_COUNT', 50);
```
FEEDBACK_SUBJECT_MAX_CHAR_LIMIT

pandeykartikey replied:
Done

seanlip commented on core/templates/dev/head/filters.js:
```diff
@@ -407,6 +409,21 @@ oppia.filter('stripFormatting', [function() {
   };
 }]);
 
+oppia.filter('getAbbreviatedText', [function () {
```
Add more tests -- very long one-word subjects, subjects without spaces, subjects with non-ascii chars. Try to cover all the weird cases that can occur without thinking about how this function is actually implemented.

pandeykartikey replied:
I have added a few more test I think those cover the cases you mentioned.

seanlip commented on core/templates/dev/head/filters.js:
```diff
@@ -407,6 +409,21 @@ oppia.filter('stripFormatting', [function() {
+    if (text.length > characterCount) {
+      var subject = text.substr(0, characterCount);
+        
+      if (subject.includes(' ')) {
```
.incudes() isn't supported in FF, I think? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/includes#compatNote_1

pandeykartikey replied:
Done

seanlip commented on core/templates/dev/head/filters.js:
```diff
@@ -407,6 +409,21 @@ oppia.filter('stripFormatting', [function() {
+        
+      if (subject.includes(' ')) {
+        subject = subject.split(' ').slice(0, -1)
+                          .join(' ');
```
Move to end of previous line.

pandeykartikey replied:
Done

seanlip reviewed (commented):
I think the flaky tests are due to #4612, which @GanitGenius is trying to address in #4631. In the meantime your best bet is to restart them, or ask a maintainer to do so --
 see [here](https://github.com/oppia/oppia/wiki/If-your-build-fails).

pandeykartikey committed:
Added more tests

seanlip commented on core/templates/dev/head/filtersSpec.js:
```diff
@@ -453,6 +454,31 @@ describe('Testing filters', function() {
+    expect($filter('getAbbreviatedText')('It will remain unchanged.', 50))
+      .toBe('It will remain unchanged.');
+    expect($filter('getAbbreviatedText')(
+      'Itisjustaverylongsinglewordfortesting', 
```
Try a word that has > 50 chars.

seanlip replied:
Oh wait, you have one below. Sorry, ignore this.

pandeykartikey replied:
I think its done in a test below that.

seanlip commented:
So I am looking at earlier comments just to make sure they've been addressed, and I have a question.

I believe the approach we decided on is the following (and Kevin agreed):

> So, here's what I suggest instead: when retrieving the list of feedback threads from the backend, do an additional database call to get the content of the first message for each of those threads, and send this info over to the frontend. The frontend takes those messages, abbreviates them, and displays them. (It's also OK to do the abbreviation in the backend if you prefer.) That way, we don't change the meaning of the "subject" field midway (otherwise we'd have some model instances that say "Feedback from a learner" and some others that have abbreviated feedback text and some others that are handwritten by fellow creators).

This isn't what you have here. Did you miss that comment?

pandeykartikey commented:
@seanlip I seem to have missed it.
I think this is kind of a hacky way. Instead, we should just update the database for the feedback title that would be a lot cleaner way.

seanlip commented:
When you say "this", what are you referring to? The implementation in this PR, or the implementation proposed in the comment?

pandeykartikey commented:
Suggested implementation.

seanlip commented:
Ah, I see. I would be OK with the approach you mention. You'd probably need to write a mapreduce job that iterates over the feedback thread models and updates their subject, and also fixes the subject line for any new threads coming in. Make sure to not change the last_updated time on those models though.

pandeykartikey commented:
Where to write one time jobs in the structure? @seanlip

seanlip commented:
I recommend talking to @pranavsid98 about this -- he should be able to help you.

pandeykartikey commented:
@pranavsid98  I have written this script can you tell me where to integrate it
```
from core.domain import feedback_domain

def fix_subject():
    feedbacks_require_fix = feedback_domain.FeedbackThread.objects.filter(subject='(Feedback from a learner)')
    for feedback in feedbacks_require_fix:
        feedback.subject = generate_new_subject(feedback.summary, 50)
        feedback.save()

def generate_new_subject(text, characterCount):
    if text.length > characterCount:
        subject = text[0:characterCount]
        
        if subject.index(' ') > -1:
            s = ' '
            subject = s.join(subject.split(' ')[:-1])
        
        return subject + '...'
    return text

if __name__ == '__main__':
    fix_subject()
```

pranavsid98 commented:
@kartik-25 One off MR jobs don't work like that. There is a specific way of writing these jobs. For easiest reference, take a look at FeedbackThreadMessagesCountOneOffJob in core.domain.feedback_jobs_one_off.

The code you've written should be easily implementable in the MR jobs. Please write a job in core.domain.feedback_jobs_one_off and then also write some tests for it. If you have any issues along the way, ping me on Hangouts or just email me!

Thanks!

1995YogeshSharma commented:
Hi @kartik-25 , how's this going? If you are facing with some issue you can ask here.

pandeykartikey commented:
@pranavsid98 @1995YogeshSharma actually I didn't understand the purpose of one-off MR job and so couldn't mould my function according to that.

1995YogeshSharma commented:
Hi @kartik-25 , MapReduce is a method used to view and update database entries in a distributed system setting. The normal database query cannot work as there are multiple servers now. So, we write a mapper and a reducer. Mapper is run on each node and then reducer combines the result.

That's a top off the head explanation. I hope it helps you appreciate the need of map reduce.

You can read any tutorial online for map reduce to get an idea. Maybe follow [this](https://hadoop.apache.org/docs/r1.2.1/mapred_tutorial.html) or the wikipedia entry for map reduce.

Reading one of the examples (as mentioned by Pranav) from the code base will then make everything clear.

pandeykartikey committed:
Adds One Off Job for feedback subject update

pandeykartikey commented:
@pranavsid98 Have a look.

pandeykartikey committed:
Fixes tests

pandeykartikey committed:
Fixes Tests

pandeykartikey commented:
@pranavsid98 I cannot understand why travis is failing because of `449:1  error  Trailing spaces not allowed  no-trailing-spaces` in `/home/travis/build/oppia/oppia/core/templates/dev/head/filters.js` When no edits are made to this line in the PR. I also tried running lint check on local and it passed without any error

apb7 commented:
Hey, @kartik-25! I'm looking into this. Don't worry :)
Can you once try deleting and adding the eof line?  (Also, the lint test has been restarted.)
Thanks!

pandeykartikey commented:
@apb7 git is not detecting any diff to commit.

apb7 commented:
Hi, @kartik-25! I figured out the problem:  

1. Your branch is 70 commits behind the oppia:develop branch. Please update it.  
2. There are trailing whitespaces on line 416 in filters.js. (I copied the raw and ran it locally.)  
3. Also, please check if there is a newline at EOF once.  

Thanks!

pandeykartikey committed:
Merge remote-tracking branch 'upstream/develop' into feedback-title

pandeykartikey committed:
Merge branch 'develop' into feedback-title

pandeykartikey commented:
@apb7 I have updated my branch but still the issue persists.

seanlip commented:
@pandeykartikey Please click on the Travis "Details" link, then click on the failing "RUN_LINT" test and read the error message. It points to the line number to fix, and describes what's wrong with it. Thanks!

pandeykartikey commented:
@seanlip The problem is that neither my local nor the diffs show the trailing spaces which the error is talking about. :frowning_face:

seanlip commented:
But the "Files Changed" tab on GitHub does. I just checked.

pandeykartikey committed:
Lint fix

pandeykartikey commented:
@pranavsid98 @1995YogeshSharma The MR jobs are done I have written some tests for them also. Have a look.

pandeykartikey commented:
@seanlip I have done the required changes have a look.

seanlip commented on core/domain/feedback_jobs_one_off.py:
```diff
@@ -66,3 +66,34 @@ def reduce(key, stringified_message_ids):
+        if len(item.summary) > characterCount:
+            updated_subject = item.summary[0:characterCount]
+
+            if updated_subject.index(' ') > -1:
```
Add a test with strings not containing a space char.

pandeykartikey replied:
Done

seanlip commented on core/domain/feedback_jobs_one_off.py:
```diff
@@ -66,3 +66,34 @@ def reduce(key, stringified_message_ids):
+            return
+        characterCount = 50
+        if len(item.summary) > characterCount:
+            updated_subject = item.summary[0:characterCount]
```
You can omit the 0.

pandeykartikey replied:
Done

seanlip commented on core/domain/feedback_jobs_one_off.py:
```diff
@@ -66,3 +66,34 @@ def reduce(key, stringified_message_ids):
+            return
+        if not item.summary:
+            return
+        characterCount = 50
```
Use UPPER_CASE for constants and consider making this a class-level constant.

We do not use camelCase for variable/constant names in Python. See the style rules.

EDIT: see also later comment re assets/constants.js

pandeykartikey replied:
Done

seanlip commented on core/domain/feedback_jobs_one_off.py:
```diff
@@ -66,3 +66,34 @@ def reduce(key, stringified_message_ids):
+            updated_subject = item.summary[0:characterCount]
+
+            if updated_subject.index(' ') > -1:
+                s = ' '
```
You don't need to declare s. Just do ' '.join(...)

pandeykartikey replied:
Done

seanlip commented on core/domain/feedback_jobs_one_off.py:
```diff
@@ -66,3 +66,34 @@ def reduce(key, stringified_message_ids):
+
+    @staticmethod
+    def map(item):
+        if item.subject != '(Feedback from a learner)':
```
Consider pulling out the hardcoded string as a class-level constant, too.

pandeykartikey replied:
Done

seanlip commented on core/domain/feedback_jobs_one_off.py:
```diff
@@ -66,3 +66,34 @@ def reduce(key, stringified_message_ids):
+            item.subject = updated_subject
+        else:
+            item.subject = item.summary
+        item.put(update_last_updated_time=False)
```
There should be a test to verify that the last updated time isn't changed.

seanlip replied:
I still don't see this test.

seanlip commented on core/domain/feedback_jobs_one_off_test.py:
```diff
@@ -125,3 +125,96 @@ def test_message_count(self):
+    EXPECTED_THREAD_DICT3 = {
+        'status': u'open',
+        'state_name': u'a_state_name',
+        'summary': u'It has to convert to a substring as it exceeds the character limit.', # pylint: disable=line-too-long
```
No need for the pragma. Do something like this:

    'summary': (
        'It has to ... exceeds '
        'the character limit.')

pandeykartikey replied:
Done

seanlip commented on core/domain/feedback_jobs_one_off_test.py:
```diff
@@ -125,3 +125,96 @@ def test_message_count(self):
+
+    def _run_one_off_job(self):
+        """Runs the one-off MapReduce job."""
+        job_id = feedback_jobs_one_off.FeedbackSubjectOneOffJob.create_new() # pylint: disable=line-too-long
```
Try and avoid the pragma if possible. Would this work?

    job_id = (
        feedback_jobs_one_off.FeedbackSubjectOneOffJob.create_new())

pandeykartikey replied:
Done

seanlip commented on core/domain/feedback_jobs_one_off_test.py:
```diff
@@ -125,3 +125,96 @@ def test_message_count(self):
+        feedback_services.create_thread(
+            self.EXP_ID_1, self.EXPECTED_THREAD_DICT1['state_name'],
+            self.user_id, self.EXPECTED_THREAD_DICT1['subject'],
+            'not used here')
```
Better to name this 'unused state name' or whatever the arg is. Ditto below.

pandeykartikey replied:
Done

seanlip commented on core/domain/feedback_jobs_one_off_test.py:
```diff
@@ -125,3 +125,96 @@ def test_message_count(self):
+            self.user_id)
+        feedback_services.add_summary_to_feedback(
+            thread_ids[0], self.EXPECTED_THREAD_DICT1['summary']
+            )
```
Move to end of previous line. Ditto below.

If you want to keep it on a separate line, deindent by 4 to match opening paren.

pandeykartikey replied:
Done

seanlip commented on core/domain/feedback_jobs_one_off_test.py:
```diff
@@ -125,3 +125,96 @@ def test_message_count(self):
+
+        threads = feedback_services.get_threads(self.EXP_ID_1)
+
+        self.assertEqual(threads[0].subject, u'a small summary')
```
Test coverage seems incomplete. How about empty strings, long strings with no spaces, etc.? Try to break your code. Otherwise there'll be lots of debugging needed in production.

pandeykartikey replied:
Added few more tests. Going to add a few more in next commit

pandeykartikey replied:
Done

seanlip commented on core/domain/feedback_jobs_one_off_test.py:
```diff
@@ -125,3 +125,96 @@ def test_message_count(self):
+                taskqueue_services.QUEUE_NAME_ONE_OFF_JOBS), 1)
+        self.process_and_flush_pending_tasks()
+
+    def test_feedback_subject(self):
```
test_that_job_returns_correct_feedback_subject

pandeykartikey replied:
Done

seanlip commented on core/domain/feedback_jobs_one_off_test.py:
```diff
@@ -125,3 +125,96 @@ def test_message_count(self):
+
+        thread_ids = subscription_services.get_all_threads_subscribed_to(
+            self.user_id)
+        feedback_services.add_summary_to_feedback(
```
I don't understand why you need this. Isn't the summary already created in create_thread()?

I'd have thought you'd just create the threads, check their summaries, run the job, re-check their summaries, and confirm that they've changed.

pandeykartikey replied:
@seanlip No, summary is not created in create_thread().
Shall I update the function to add summary (by default `null`)?

seanlip replied:
Oh, I see, you're still preparing the threads. In that case why can't you use existing functions to do this?

(I mean, how does the thread even get a summary at all during normal operation, currently? Adding a new function just for tests is still not a great idea.)

seanlip commented on core/domain/feedback_services.py:
```diff
@@ -94,6 +94,18 @@ def create_thread(
         exploration_id, state_name, original_author_id, subject, text, False)
 
 
+def add_summary_to_feedback(thread_id, summary):
```
Avoid creating entire functions in main code that are used only by tests. That's an antipattern.

seanlip commented on core/templates/dev/head/filters.js:
```diff
@@ -18,6 +18,8 @@
 
 oppia.constant('RULE_SUMMARY_WRAP_CHARACTER_COUNT', 30);
 
+oppia.constant('FEEDBACK_SUBJECT_MAX_CHAR_LIMIT', 50);
```
Use assets/constants.js for constants that are shared between frontend and backend. The backend should read from this as well so that there's no skew.

pandeykartikey replied:
Done

seanlip commented on core/templates/dev/head/filters.js:
```diff
@@ -465,6 +467,20 @@ oppia.filter('stripFormatting', [function() {
 }]);
 
+oppia.filter('getAbbreviatedText', [function () {
+  return function (text, characterCount) {
```
Delete space after `function`

pandeykartikey replied:
Done

seanlip commented on core/templates/dev/head/filters.js:
```diff
@@ -465,6 +467,20 @@ oppia.filter('stripFormatting', [function() {
+    if (text.length > characterCount) {
+      var subject = text.substr(0, characterCount);
+
+      if (subject.indexOf(' ') > -1) {
```
Consider !== rather than >. They do the same thing but conceptually -1 is a sentinel value here rather than "one position less than 0".

pandeykartikey replied:
Done

seanlip reviewed (changes requested):
Hi @pandeykartikey, thanks! On the whole I think this looks good but the backend needs a bit more work.

pandeykartikey committed:
Made the Required Changes

pandeykartikey committed:
Adds more test

seanlip commented on core/domain/feedback_jobs_one_off.py:
```diff
@@ -71,23 +72,26 @@ def reduce(key, stringified_message_ids):
+        if len(item.summary) > constants.FEEDBACK_SUBJECT_MAX_CHAR_LIMIT:
+            updated_subject = item.summary[
+                :constants.FEEDBACK_SUBJECT_MAX_CHAR_LIMIT
+                ]
```
Either deindent this by 4 or move it to the end of the previous line.

seanlip commented on core/domain/feedback_jobs_one_off_test.py:
```diff
@@ -169,46 +189,54 @@ def setUp(self):
-            self.EXP_ID_1, self.EXPECTED_THREAD_DICT1['state_name'],
+            self.EXP_ID_1, self.EXPECTED_THREAD_DICT1['unused_state_name'],
             self.user_id, self.EXPECTED_THREAD_DICT1['subject'],
             'not used here')
```
Sorry, state_name was fine as-is. I was referring to the "not used here" string, which can presumably be replaced with "unused something".

seanlip commented on scripts/run_backend_tests.sh:
```diff
@@ -88,7 +88,7 @@ for arg in "$@"; do
 done
 
-$PYTHON_CMD scripts/backend_tests.py $@
+$PYTHON_CMD scripts/backend_tests.py --test_target core.domain.feedback_jobs_one_off_test $@
```
Check the PR diff before you push. You can use tools like `meld`.

seanlip commented on core/templates/dev/head/filters.js:
```diff
@@ -18,7 +18,8 @@
 oppia.constant('RULE_SUMMARY_WRAP_CHARACTER_COUNT', 30);
 
-oppia.constant('FEEDBACK_SUBJECT_MAX_CHAR_LIMIT', 50);
+oppia.constant('FEEDBACK_SUBJECT_MAX_CHAR_LIMIT',
```
Break after '(' if the contents of the parens span multiple lines.

seanlip reviewed (changes requested):
Hi @pandeykartikey, I've taken another pass!

One request: in the next review round, to make your comments, could you please go to "Files changed", click "start a review" with your first comment, enter your line comments, and then click "Review changes" and "Submit review" using the green button at the top-right? That will help batch the conversation history into meaningful chunks. Thanks!

pandeykartikey committed:
Fixes

pandeykartikey commented on core/domain/feedback_jobs_one_off_test.py:
```diff
@@ -125,3 +125,120 @@ def test_message_count(self):
+        self.assertEqual(
+            threads[5].subject,
+            u'Itisjustaverylongsinglewordfortesting')
+        self.assertEqual(threads[0].last_updated, threads_old[0].last_updated)
```
Added the test for last_updated

pandeykartikey commented on core/domain/feedback_jobs_one_off_test.py:
```diff
@@ -125,3 +125,120 @@ def test_message_count(self):
+    def test_that_job_returns_correct_feedback_subject(self):
+        """Test if the job returns the correct feedback subject."""
+        feedback_services.create_thread(
+            self.EXP_ID_1, 'unused_state_name', self.user_id,
```
Moved unused_state_name out of the `dict`.

pandeykartikey commented on core/templates/dev/head/filters.js:
```diff
@@ -18,6 +18,10 @@
 oppia.constant('RULE_SUMMARY_WRAP_CHARACTER_COUNT', 30);
 
+oppia.constant(
+  'FEEDBACK_SUBJECT_MAX_CHAR_LIMIT',
```
Moved it to next line

pandeykartikey reviewed (commented):
@seanlip Actually first I had made a mistake first message was supposed to be used and not summary. I have fixed that. Have a look.

seanlip commented on core/domain/feedback_jobs_one_off.py:
```diff
@@ -72,32 +72,35 @@ def reduce(key, stringified_message_ids):
+        yield (item.id, item.subject)
+
+    @staticmethod
+    def reduce(key, stringified_subject):
```
Why are you doing all this stuff in reduce? Wouldn't it be better to do it in map instead? That way it doesn't have to go through the shuffle stage.

seanlip commented on core/domain/feedback_jobs_one_off_test.py:
```diff
@@ -132,45 +132,39 @@ class FeedbackSubjectOneOffJobTest(test_utils.GenericTestBase):
         'subject': u'(Feedback from a learner)'
     }
 
-    EXPECTED_THREAD_DICT7 = {
```
Why delete? Weird characters in the text is a case to test too.

pandeykartikey committed:
Made the required changes

pandeykartikey commented on core/domain/feedback_jobs_one_off_test.py:
```diff
@@ -125,3 +125,133 @@ def test_message_count(self):
+        'subject': u'(Feedback from a learner)'
+    }
+
+    EXPECTED_THREAD_DICT7 = {
```
Added the case for weird characters.

pandeykartikey commented on core/domain/feedback_jobs_one_off.py:
```diff
@@ -66,3 +67,40 @@ def reduce(key, stringified_message_ids):
+    @staticmethod
+    def map(item):
+        if item.subject != FeedbackSubjectOneOffJob.DEFAULT_SUBJECT:
+            return
```
Moved it back to map function.

pandeykartikey reviewed (commented):
@seanlip Made the required changes have a look.

seanlip commented on core/domain/feedback_jobs_one_off.py:
```diff
@@ -66,3 +67,40 @@ def reduce(key, stringified_message_ids):
+
+            if ' ' in updated_subject:
+                updated_subject = ' '.join(updated_subject.split(' ')[:-1])
+            updated_subject = updated_subject + '...'
```
Nit: I don't think you need to redefine updated_subject. Just replace this line and the next with `item.subject = updated_subject + '...'`.

seanlip commented on core/domain/feedback_jobs_one_off_test.py:
```diff
@@ -125,3 +125,133 @@ def test_message_count(self):
+            self.EXP_ID_1, self.owner_id, title='Bridges in England',
+            category='Architecture', language_code='en')
+
+
```
Drop one of the newlines.

pandeykartikey committed:
Minor fixes

pandeykartikey commented:
@seanlip made the required changes.

seanlip commented:
@pandeykartikey I'm trying to test this and I can't find a way to run the job. Did you test this with existing data? That's an important flow.

Basically:

1. Checkout the develop branch, create and add some feedback threads.
2. Checkout your feature branch.
3. Run the one-off job (from the Jobs tab in the Admin page).
4. Add more feedback threads.
5. Check that everything works correctly.

I can't do step 3. Try and make sure you've tested this fully, otherwise we'll run into issues in production.

Thanks!

pandeykartikey committed:
Adds job to registry

pandeykartikey commented:
@seanlip I have added the job to registry and verified that it is working.

seanlip reviewed (approved):
All good. Thanks!
